pkg/regserver/ampCacheregserver/ampCacheregserver.go

```diff
@@ -0,0 +1,265 @@
+package ampCacheregserver
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/gorilla/mux"
+	"github.com/refraction-networking/conjure/pkg/metrics"
+	"github.com/refraction-networking/conjure/pkg/phantoms"
+	"github.com/refraction-networking/conjure/pkg/regserver/regprocessor"
+	pb "github.com/refraction-networking/conjure/proto"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
+	"gitlab.torproject.org/tpo/anti-censorship/pluggable-transports/snowflake/v2/common/amp"
+	"google.golang.org/protobuf/proto"
+)
+
+type registrar interface {
+	RegisterUnidirectional(*pb.C2SWrapper, pb.RegistrationSource, []byte) error
+	RegisterBidirectional(*pb.C2SWrapper, pb.RegistrationSource, []byte) (*pb.RegistrationResponse, error)
+}
+
+type AMPCacheRegServer struct {
+	apiPort          uint16
+	ampCacheURL      string
+	latestClientConf *pb.ClientConf // Latest clientConf for sharing over RegistrationResponse channel.
+	ccMutex          sync.RWMutex
+	processor        registrar
+	logger           log.FieldLogger
+	logClientIP      bool
+	metrics          *metrics.Metrics
+}
+
+func (s *AMPCacheRegServer) getC2SFromReq(w http.ResponseWriter, r *http.Request, path string) (*pb.C2SWrapper, error) {
+	const MinimumRequestLength = regprocessor.SecretLength + 1 // shared_secret + VSP
+	if r.Method != "GET" {
+		s.logger.Errorf("rejecting request due to incorrect method %s\n", r.Method)
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return nil, errors.New("incorrect method")
+	}
+
+	if r.ContentLength < MinimumRequestLength {
+		s.logger.Errorf("rejecting request due to short content-length of %d, expecting at least %d\n", r.ContentLength, MinimumRequestLength)
+		http.Error(w, "Payload too small", http.StatusBadRequest)
+		return nil, errors.New("payload too small")
+	}
```

Comment:
Our registration request won't be in the body of the GET request anymore, it will be in the path, so this check should be moved to after we decode the path and check the length of the decoded payload.